Functions to split/combine multiple layers along an axis

cwood1967 opened the pull request:
# Description
This pull request adds functions to split image layers along an axis and combine image layers into a new single layer along a new axis as described in issue #1103.

This feature will make it easy to split multichannel images that are opened into a single layer, and allow split images to be recombined. Also some print statements when input layers don't make sense; not sure how to handle that.

I put these functions in napari/utils/axis_functions.py; I wasn't sure the best place for it.

I noticed that creating and deleting layers of different dimensions will throw errors that seem related to viewer.dims not being up to date with viewer.active_layer.dim, but this happens just by adding and deleting without these functions.  I added issue #1323 for this.

## Type of change
- [ ] New feature (non-breaking change which adds functionality)
- [ ] This change requires a documentation update

# References
This is for feature issue #1103.


# How has this been tested?
Manually tested.

## Final checklist:
- [ ] My PR is the minimum possible work for the desired functionality
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works

cwood1967 committed:
split and combine axis

cwood1967 committed:
Merge branch 'master' into ChrisWood
merging with napari upstream

cwood1967 committed:
moved axis_functions to napari/utils

cwood1967 committed:
Merge branch 'master' into ChrisWood

cwood1967 committed:
rename split_axis to stack_to_images. Make suggested changes

cwood1967 committed:
rename combine_layers to images_to_stack. Made suggested changes

cwood1967 committed:
moved napari/utils/axis_functions.py -> napari/layers/utils/stack_utils.py

cwood1967 committed:
removed Image from imports, moved functions to layer_utils

GenevieveBuckley commented on napari/layers/utils/layer_utils.py:
```diff
@@ -233,3 +234,101 @@ def compute_multiscale_level(
+            shape = images[i].data.shape
+        else:
+            if shape != images[i].data.shape:
+                print("Layers not the same size")
```
I think it would be more efficient to not have this check every time we go through the loop, but instead rely on `np.stack` to throw an appropriate error. Then we can use a try/except around `np.stack` to handle that.

cwood1967 replied:
Yes, that is much better. I issued a warning if a ValueError is raised.

GenevieveBuckley commented on napari/layers/utils/layer_utils.py:
```diff
@@ -233,3 +234,101 @@ def compute_multiscale_level(
+
+    new_list = list()
+    for i in range(nc):
+        # if not isinstance(images[i], Image):
```
Could you delete the commented lines?

GenevieveBuckley commented on napari/layers/utils/layer_utils.py:
```diff
@@ -233,3 +234,101 @@ def compute_multiscale_level(
+        List of images
+    """
+
+    # if not isinstance(stack, Image):
```
Could you delete the commented lines? 

We should use type hinting in the function signature instead of these kinds of checks. Note that both `Image` and `Labels` layers should be accepted by these functions (label layers are more or less a special case of images in napari)

cwood1967 replied:
I added the type hinting to both functions. Is the hinting used at runtime anywhere?

GenevieveBuckley commented on napari/layers/utils/layer_utils.py:
```diff
@@ -2,6 +2,7 @@
 
 import numpy as np
 from vispy.color import Colormap
+from napari.utils.colormaps import simple_colormaps
```
I'm uneasy about the idea of automatically choosing colormaps for users. Ultimately I think this problem is too open-ended for us to reliably make a good and useful default choice. I think the default should be "you keep the colormap you had originally in the image stack"

I would be happy to see colormaps as an optional keyword argument, so users could pass in a list of colormaps in the order that they want them. (We could maybe do that with all possible keyword arguments too, if we wanted)

sofroniewn replied:
I'd been thinking about this too. One approach would be to use `layer._get_state()` to get and propagate as many possible properties as possible. I do like letting people pass lists of properties too. I'd try and mirror the approach of `add_image` as much as possible. There we have sensible defaults if two, three, or more images are being stacked

cwood1967 replied:
I agree about the colormaps. The keyword argument was added to work with strings or vispy Colormaps.

GenevieveBuckley commented on napari/layers/utils/layer_utils.py:
```diff
@@ -233,3 +234,101 @@ def compute_multiscale_level(
+
+    nc = len(images)
+    if nc < 2:
+        print("less than two layers selected")
```
I don't think this check is needed, to be honest. Nothing bad happens if you try and stack a single image together with nothing else, so I don't think we need to waste energy trying to prevent people from trying to do it (accidentally or no)

GenevieveBuckley commented on napari/layers/utils/layer_utils.py:
```diff
@@ -233,3 +234,101 @@ def compute_multiscale_level(
+        return None
+
+    new_list = list()
+    for i in range(nc):
```
This is simpler if we write it as 
```python
data = []
for image in images:
    data.append(image.data)
    ...
```

GenevieveBuckley commented on napari/layers/utils/layer_utils.py:
```diff
@@ -233,3 +234,101 @@ def compute_multiscale_level(
+
+    new_data = np.stack(new_list, axis=axis)
+
+    stack = type(images[0])(new_data)
```
Oh I see what's happened here, this must be the spot where circular imports were making things tricky. @sofroniewn should we shift all of this someplace else?

sofroniewn replied:
Would it help if things were in our https://github.com/napari/napari/blob/master/napari/layers/image/_image_utils.py?

cwood1967 replied:
I tried using _image_utils.py, but it looks like _image_utils.py is imported when Image is imported:
```~/Code/napari/napari/layers/image/__init__.py in <module>
----> 1 from .image import Image

~/Code/napari/napari/layers/image/image.py in <module>
     11 from ..intensity_mixin import IntensityVisualizationMixin
     12 from ._image_constants import Interpolation, Interpolation3D, Rendering
---> 13 from ._image_utils import guess_rgb, guess_multiscale
     14
     15

~/Code/napari/napari/layers/image/_image_utils.py in <module>
      3
      4 import numpy as np
----> 5 from napari.layers.image.image import Image
      6 from napari.layers.labels.labels import Labels
      7 from vispy.color import Colormap
```

GenevieveBuckley commented on napari/layers/utils/layer_utils.py:
```diff
@@ -233,3 +234,101 @@ def compute_multiscale_level(
+    num_dim = len(data.shape)
+
+    if num_dim < 3:
+        print("not enough dimensions")
```
A python warning might be more appropriate than a print statement for this sort of thing

https://docs.python.org/3/library/warnings.html

GenevieveBuckley reviewed (commented):
I've left a few more comments.

You're right, circular imports are definitely a problem (sorry I didn't realize the point you were making earlier!). I'd like @sofroniewn to suggest a better place for these functions to live

cwood1967 committed:
added type hints, updated how colormaps are assigned, moved functions from layer_utils to stack_utils

cwood1967 committed:
stack_to_images colormap defaults to input layer colorm map

jni commented on napari/layers/utils/stack_utils.py:
```diff
@@ -0,0 +1,103 @@
+
+    cmap = stack.colormap
+    if colormaps is None:
+        colormaps = data.shape[axis] * [cmap]
```
Sorry to restart this discussion, but I think when splitting a stack, we should follow the convention of `add_image(..., channel_axis=ax)`. This is handled here:

https://github.com/napari/napari/blob/b94ed99b8f60599aba61579f5711c8bf6e8896d2/napari/components/add_layers_mixin.py#L236-L247

We can definitely argue that choice and change it if we decide it was the wrong choice, but we haven't had any complaints, and either way the behaviours should match, preferably even through shared code.

jni commented on napari/layers/utils/stack_utils.py:
```diff
@@ -0,0 +1,103 @@
+    imagelist = list()
+
+    for i in range(data.shape[axis]):
+        layer_name = "{:02d}_{}".format(i, name)
```
@cwood1967 although I understand the desire to keep the channel information out front, it really doesn't gel with convention. Ultimately, I think this is a usability fix that needs to happen on how we display layer names, but in my opinion the image number should appear at the end. I'd propose here:

```suggestion
        layer_name = f'{name} layer {i}'
```

or

```python
        layer_name = f'{name} ch{i}'
```

for example.

jni commented on napari/layers/utils/stack_utils.py:
```diff
@@ -0,0 +1,103 @@
+
+
+def images_to_stack(
+    images: List[Union[Image, Labels]], axis: int = -1
```
It would be much more efficient to combine along `axis=0` by default... Not to mention, it would be guaranteed have the desired effect of adding a slider to the viewer, whereas the converse might result in rgb display (if there are three channels) or a super-thin slice being displayed from each channel (if there is any other number of channels).

jni commented on napari/layers/utils/stack_utils.py:
```diff
@@ -0,0 +1,103 @@
+
+    stack = Image(new_data)
+
+    return stack
```
I don't think we should necessarily be in the business of catching this error, vs letting it bubble up to the caller. If we do the latter, the function becomes super simple:

```python
new_data = np.stack([image.data for image in images], axis=axis)
stack = Image(new_data)
return stack
```

We might want to propagate things such as scale and translation into the new Image, though...

jni reviewed (commented):
@cwood1967 thanks for this! I've made some comments proposing a few changes for discussion. Additionally, some of the test scripts you've proposed in comments should be added as actual tests before merging.

The main thing that I think should be implemented within these functions is propagating other settings from the layers, as proposed by @sofroniewn. Several things such as scale and translation should be transferable with a bit of processing. (ie inserting a 1 in the position of the new axis, or removing the value in the position of the removed axis.)

cwood1967 committed:
remove try/except from images_to_stack

cwood1967 committed:
put layer number at the end of the name

cwood1967 committed:
added tests in napari/layers/utils/_tests/test_stack_utils.py

cwood1967 committed:
colormap assignment like add_image

cwood1967 committed:
propagating settings

cwood1967 committed:
Merge branch 'master' into ChrisWood

cwood1967 committed:
propagate settings and add kwargs

cwood1967 committed:
colormaps to colormap in test_stack_utils

cwood1967 committed:
convert colormap list to iter

cwood1967 committed:
changed grey to gray

cwood1967 committed:
Merge branch uptream 'master' into ChrisWood

cwood1967 committed:
added parameters and kwargs to images_to_stack

jni commented on napari/layers/utils/stack_utils.py:
```diff
@@ -0,0 +1,171 @@
+        if n_channels == 2:
+            colormap = iter(colormaps.MAGENTA_GREEN)
+        if n_channels > 2:
+            colormap = itertools.cycle(colormaps.CYMRGB)
```
I think there should be an else clause here, for rare cases when an axis might be a singleton.

```suggestion
    if colormap is None:
        if n_channels == 2:
            colormap = iter(colormaps.MAGENTA_GREEN)
        elif n_channels > 2:
            colormap = itertools.cycle(colormaps.CYMRGB)
        else:
            colormap = itertools.repeat('gray')
```

jni commented on napari/layers/utils/stack_utils.py:
```diff
@@ -0,0 +1,171 @@
+        if n_channels > 2:
+            colormap = itertools.cycle(colormaps.CYMRGB)
+    else:
+        colormap = iter(colormap)
```
I also think we should cycle here for times when you have more channels than colours:

```suggestion
        colormap = itertools.cycle(colormap)
```

jni commented on napari/layers/utils/stack_utils.py:
```diff
@@ -0,0 +1,171 @@
+        blending = 'additive'
+
+    if scale is None:
+        scale = np.delete(stack.scale, axis)
```
Whoa. Didn't know about `np.delete`. Nice use of it here!

jni commented on napari/layers/utils/stack_utils.py:
```diff
@@ -0,0 +1,171 @@
+        try:
+            color = next(colormap)
+        except IndexError:
+            color = 'gray'
```
I think this try/except can be removed if you implement my suggestions from above.

jni replied:
And actually I think this was incorrect, as the above will raise a StopIteration error. Whatever path we decide to adopt, this should be tested!

jni commented on napari/layers/utils/stack_utils.py:
```diff
@@ -0,0 +1,171 @@
+
+        image = Image(np.take(data, i, axis=axis), name=layer_name, **kwargs)
+
+        imagelist.append(image)
```
@cwood1967 you're right that this is now very similar to `add_image, axis=not_none`, in fact, maybe a little too similar? I feel like the two code chunks should match nearly exactly, and possibly even factored out into a single approach. Here's the code for reference:

https://github.com/napari/napari/blob/e43f730170aab1e011c42060c65ba44133f6f18f/napari/components/add_layers_mixin.py#L226-L275

tlambert03 replied:
yeah, this is my main concern here too, this is pretty much the same thing as `channel_axis`.  So as @jni suggested, my feeling is that we should pull out the axis splitting logic from `viewer.add_image()` into a helper function that would also take the place of this `stack_to_images` function here, and `viewer.add_images` should call that function. 
 thoughts on that @jni @sofroniewn, @GenevieveBuckley?

The signature of that function could look like this
```python
def split_channels(data: ndarray, axis: int, **kwargs) -> List[LayerData]:
    ...
```

and then the `if channel_axis is not None` clause in `viewer.add_image` could look like:

```python
for datum in split_channels(data, **kwargs):
    self._add_layer_from_data(**datum)
```

and then splitting an image that's already been added to the viewer would be something like:

```python
def split_image(self, image: Image, axis, **kwargs):
    data, meta, layer_type = image.as_layer_data_tuple()
    if not meta["rgb"]:
        meta['scale'].pop(axis)
        meta['translate'].pop(axis)
    for key in ("contrast_limits", "colormap", "blending", "rgb"):
        del meta[key]
    meta.update(kwargs)
    self.layers.remove(image)
    self.add_image(data, channel_axis=axis, **meta)
    # or call the split_channels() directly
```

@cwood1967, regardless of which direction we go there... a good thing to point out in this function is the `data, meta, layer_type = image.as_layer_data_tuple()` which will let you remove a lot of code from your function where you gather up the metadata of the current image.

sofroniewn replied:
This change has been made I think

jni commented on napari/layers/utils/stack_utils.py:
```diff
@@ -0,0 +1,171 @@
+    }
+    new_list = [image.data for image in images]
+    new_data = np.stack(new_list, axis=axis)
+    stack = Image(new_data, **kwargs)
```
This one still needs some work. It's unclear what should be done with things like colormap and contrast limits, but scale and translate should certainly be pilfered from the top image and a scale of `1` inserted at the new axis.

jni reviewed (changes requested):
@cwood1967 sorry for the slow-burn reviewing here and thank you for your patience! The good thing is that the code is nice, compact, and clean, which made a full review very easy. My main issues are:

- the `stack_to_images` function needs to be reworked a little to prevent (different!) code duplication with `AddLayersMixin.add_image`, and
- the `images_to_stack` function should preserve scale and translate parameters, even if it discards the others.

The tests should also exercise a few more code paths.

It's very close though! Thanks!

cwood1967 committed:
account for rgb

cwood1967 committed:
Merge branch 'master' into ChrisWood

cwood1967 committed:
removed if __name__ from tests

tlambert03 commented on napari/layers/utils/_tests/test_stack_utils.py:
```diff
@@ -0,0 +1,124 @@
+
+    '''
+    Test 2 channels, no colormap
+    '''
```
```suggestion
    """Test 2 channels, no colormap."""
```
([PEP 257](https://www.python.org/dev/peps/pep-0257/))

Also, if possible, I think it's nice to say what expected behavior you're testing.  like
```python
"""Test that a multichannel stack is split into single-channel layers."""
```

tlambert03 commented on napari/layers/utils/_tests/test_stack_utils.py:
```diff
@@ -0,0 +1,124 @@
+    '''
+    Test 3 channels, last index, no colormap - RGB
+    from rgb to single channels
+    '''
```
```suggestion
    """Test 3 channels, negative index, no colormap - RGB to single channels."""
```

tlambert03 commented on napari/layers/utils/_tests/test_stack_utils.py:
```diff
@@ -0,0 +1,124 @@
+def test_stack_to_images_4_channels():
+    '''
+    Test 4 channels, first index, colormap
+    '''
```
```suggestion
    """Test 4 channels, first index, colormap."""
```

tlambert03 commented on napari/layers/utils/stack_utils.py:
```diff
@@ -0,0 +1,198 @@
+    blending: str = None,
+    scale: Tuple[float] = None,
+    translate: Tuple[float] = None,
+) -> List[Image]:
```
thank you for the type annotations!! 😍

tlambert03 commented on napari/layers/utils/_tests/test_stack_utils.py:
```diff
@@ -0,0 +1,124 @@
+from napari.layers.image.image import Image
```
```suggestion
from napari.layers import Image
```

tlambert03 commented on napari/layers/utils/stack_utils.py:
```diff
@@ -0,0 +1,198 @@
+
+
+def stack_to_images(
+    stack: Union[Image, Labels],
```
```suggestion
    stack: Image,
```
(`Labels` _is_ an `Image`... so this is unnecessary)

tlambert03 reviewed (commented):
left some thoughts.  I agree with @jni that this is too much like `viewer.add_image` with `channel_axis`... so we should probably extract that into a more reusable function.  I left some strategies for doing so

cwood1967 committed:
formatted comments in tests

cwood1967 committed:
made StackUtils class and added split_channels

cwood1967 committed:
remove kwargs contruction from image_to_stack and us as_layer_data_tuple()

cwood1967 committed:
Removed splitting channels from AddLayersMixin and called
split_channels from StackUtils

cwood1967 committed:
Removed StackUtils class and just used functions

cwood1967 committed:
updated doc strings

jni reviewed (approved):
I don't have time for a deep review right now, but after a quick glance this looks great! Super nice structure now.

Some of the docstrings need a bit of fleshing out (e.g. missing description of "data") but I don't think it should hold up the PR.

Thanks @cwood1967!

sofroniewn commented on napari/components/add_layers_mixin.py:
```diff
@@ -229,55 +224,13 @@ def add_image(
-                            f"({channel_axis}) had length {n_channels}, but "
-                            f"the '{key}' argument only provided {i} values. "
-                        )
+            layerdata_list = split_channels(data, channel_axis, **kwargs)
```
Wow, love being able to use `split_channels` here, very nice!

sofroniewn commented on napari/layers/utils/_tests/test_stack_utils.py:
```diff
@@ -0,0 +1,174 @@
+
+
+def test_stack_to_images_basic():
+
```
```suggestion
```

sofroniewn commented on napari/layers/utils/stack_utils.py:
```diff
@@ -0,0 +1,188 @@
+    multiscale = kwargs['multiscale']
+    if multiscale is None:
+        multiscale, data = guess_multiscale(data)
+    kwargs['multiscale'] = multiscale
```
```suggestion
    if 'multiscale' not in kwargs or kwargs['multiscale'] is None:
        multiscale, data = guess_multiscale(data)
        kwargs['multiscale'] = multiscale
```

sofroniewn replied:
I'm suggesting this change so code will still run even if `multiscale` not in kwargs

sofroniewn commented on napari/layers/utils/stack_utils.py:
```diff
@@ -0,0 +1,188 @@
+        multiscale, data = guess_multiscale(data)
+    kwargs['multiscale'] = multiscale
+    n_channels = (data[0] if multiscale else data).shape[channel_axis]
+    kwargs['blending'] = kwargs['blending'] or 'additive'
```
```suggestion
    if 'blending' not in kwargs or kwargs['blending'] is None:
        kwargs['blending'] = 'additive'
```
Suggesting this change so code will run even if 'blending' is not in kwargs

sofroniewn commented on napari/layers/utils/stack_utils.py:
```diff
@@ -0,0 +1,188 @@
+    Returns
+    -------
+    List of LayerData tuples: [(data: array, metadata: Dict, type: str )]
+
```
```suggestion
```

sofroniewn commented on napari/layers/utils/stack_utils.py:
```diff
@@ -0,0 +1,188 @@
+
+    name = stack.name
+    num_dim = len(data.shape)
+    # n_channels = data.shape[axis]
```
```suggestion
```

sofroniewn commented on napari/layers/utils/stack_utils.py:
```diff
@@ -0,0 +1,188 @@
+    images: List[Union[Image, Labels]], axis: int = 0, **kwargs: Dict,
+) -> Image:
+    """Function to combine selected image layers in one layer
+
```
```suggestion

    The new image layer will get the meta data properties of the first image layer in the stack.
```

sofroniewn commented on napari/layers/utils/stack_utils.py:
```diff
@@ -0,0 +1,188 @@
+
+
+def images_to_stack(
+    images: List[Union[Image, Labels]], axis: int = 0, **kwargs: Dict,
```
I guess I see the value in allowing merging `Images` and `Labels` together, but we might need to be more careful inside this function then - for example right now if the first layer is in the list is a labels then our assumptions about propagating metadata will be off.

I think for this PR I'd be fine dropping the `Labels` from the function signature, keeping the assumption of only images and then revisiting this as part of #1442 where we want to add more functionality to convert between images and labels anyway.

tlambert03 replied:
if you mean dropping the type annotation from the function signature, it A) doesn't change anything about the runtime (as you know) but B) it actually doesn't even change anything about the types! because Labels is an Image.  That is, as far as types are concerned `Union[Image, Labels]` is the same thing as `Image` (but as far as the reader of the signature is concerned ... it's true they might think that this works for labels, which we should at least test if we are going to say that)

sofroniewn commented on napari/layers/utils/stack_utils.py:
```diff
@@ -0,0 +1,188 @@
+        Combined image stack
+    """
+
+    data, meta, _ = images[0].as_layer_data_tuple()
```
```suggestion
    if len(images) > 0:
        data, meta, _ = images[0].as_layer_data_tuple()
    else:
        return None
```
Curious what people think we should do if the list is empty, I suggest returning `None`, @tlambert03 @jni?

tlambert03 replied:
I think it should raise an exception.  If we return `None`, then the return type should be changed to `Optional[Layer]`, which reminds you that any consuming function needs to make sure it got a response.

sofroniewn reviewed (approved):
This is looking great @cwood1967! I'm approving now too, I've gone through and left some suggestions, most are fairly minor. @tlambert03 might want to have a quick scan through my suggestions too, not that I'm pointing any of this in the wrong direction. I love how we're now using this functionality internally and have pulled all this logic out of `add_image` itself!!

tlambert03 commented on napari/layers/utils/stack_utils.py:
```diff
@@ -0,0 +1,188 @@
+from napari.layers import Image, Labels
+from napari.utils.colormaps import colormaps
+from napari.layers.image._image_utils import guess_multiscale
+from napari.utils.misc import ensure_iterable, ensure_sequence_of_iterables
```
we've been generally sticking with relative imports, except in tests where we use absolute imports

tlambert03 commented on napari/layers/utils/stack_utils.py:
```diff
@@ -0,0 +1,188 @@
+
+def split_channels(
+    data: np.ndarray, channel_axis: int, **kwargs,
+) -> List[Tuple]:
```
```suggestion
) -> List[FullLayerData]:
```
and then add `from ...types import FullLayerData`

tlambert03 commented on napari/layers/utils/stack_utils.py:
```diff
@@ -0,0 +1,188 @@
+    """
+
+    # Determine if data is a multiscale
+    multiscale = kwargs['multiscale']
```
even though the main "consumer" of this function is currently `viewer.add_image`, which _will_ have 'multiscale' in the `kwargs` dict... to make this more generally usable as a function that can split an array into multiple arrays, we shouldn't assume anything about the `kwargs` parameter.  In other words, the function should execute without error even if kwargs is an empty dict `{}`.  This line and other direct `kwarg[]` access lines will through a `KeyError` if the key doesn't exist.  Use `kwargs.get("key")` instead, which will return `None` if the key doesn't exist

tlambert03 replied:
sorry, just saw @sofroniewn comments to the same effect.  I agree.  But instead of 
```python
if 'multiscale' not in kwargs or kwargs['multiscale'] is None:
```
use
```python
if not kwargs.get("multiscale"):
```

tlambert03 commented on napari/layers/utils/stack_utils.py:
```diff
@@ -0,0 +1,188 @@
+    data : array or list of array
+    channel_axis : int
+        Axis to split the image along.
+    kwargs: keyword arguments
```
expand on the explanation of `kwargs`: something like "keyword arguments will override the keys in the ``meta`` dict of each returned layer data tuple".

tlambert03 commented on napari/layers/utils/stack_utils.py:
```diff
@@ -0,0 +1,188 @@
+                )
+
+        layerdata = (image, i_kwargs, 'image')
+        layerdata_list.append(layerdata)
```
@jni, @sofroniewn, should we make this function a generator that yields layer data?

sofroniewn replied:
I defer to @jni

jni replied:
I’m happy with a list here. We don’t expect this to be infinite or too big for RAM or anything like that, and lists are easier to work with.

tlambert03 commented on napari/layers/utils/stack_utils.py:
```diff
@@ -0,0 +1,188 @@
+    data: np.ndarray, channel_axis: int, **kwargs,
+) -> List[Tuple]:
+    """Function to split the data array into seperate arrays along an axis
+
```
This function now deserves some more explanation.  For instance, explain that this function will alter the blending mode, and the colormap, unless specifically overridden. etc...  (I know we didn't have that in the `add_image` docstring, but now that we have a dedicated function for it, we should explain what we're doing

tlambert03 commented on napari/layers/utils/stack_utils.py:
```diff
@@ -0,0 +1,188 @@
+    data, meta, _ = stack.as_layer_data_tuple()
+
+    for key in ("contrast_limits", "colormap", "blending"):
+        meta[key] = None
```
maybe `del meta[key]` instead of `= None`? This way, if in the future we change the signature of the `Image()` to no longer allow `None`, this will still work.

tlambert03 commented on napari/layers/utils/stack_utils.py:
```diff
@@ -0,0 +1,188 @@
+
+
+def stack_to_images(stack: Image, axis: int, **kwargs: Dict,) -> List[Image]:
+    """Function to split the active layer into separate layers along an axis
```
You can leave out the "Function to" part from all the docstrings

tlambert03 replied:
also, this function has no concept of an "active" or "inactive" layer.  So the docstring should be more specific.  "Splits a single `Image` layer into a list of Image layers along `axis`"

tlambert03 commented on napari/layers/utils/stack_utils.py:
```diff
@@ -0,0 +1,188 @@
+            ),
+            UserWarning,
+        )
+        return None
```
I think these should raise exceptions rather than warnings.  Otherwise the return type is incorrect and should be changed.  Alternatively, remove this altogether and let the `split_channels_function` below raise the exception (which it will if they are out of bounds)

tlambert03 reviewed (commented):
great progress.  I think this is generally the right organization now.  I have a handful of comments about the implementation still though.  Ping me when ready for another look

cwood1967 committed:
changed napari imports to relative

cwood1967 committed:
Updates after review:
- changed to napari relative imports
- raised exception in images_to_stack if images list is empty
- raised exceptions rather than warnings in stack_to_images
- updated doc strings
- using dict.get() to avoid key errors

cwood1967 committed:
added multiscale test, updated doc strings

cwood1967 committed:
have split_channels always assign colormaps

cwood1967 committed:
fixed doc string

tlambert03 commented on napari/layers/utils/stack_utils.py:
```diff
@@ -49,7 +49,7 @@ def split_channels(
 
     kwargs['blending'] = kwargs.get('blending') or 'additive'
-
+    kwargs['colormap'] = kwargs.get('colormap')
```
```suggestion
    kwargs.setdefault('colormap', None)
```
handy trick ;)

tlambert03 committed:
clean tests

tlambert03 committed:
minor cleanup

tlambert03 reviewed (approved):
This looks great @cwood1967!  Thanks for the hard work on this.  I'll approve and leave it for @jni if he wants to take a last look.
Note, I pushed a couple small changes to the tests and syntax here and there, so be sure to pull first if you make additional changes.

cwood1967 committed:
change back to kwargs["blending"] =  rather than kwargs.setdefault

sofroniewn commented on napari/layers/utils/stack_utils.py:
```diff
@@ -0,0 +1,212 @@
+
+    n_channels = (data[0] if multiscale else data).shape[channel_axis]
+
+    kwargs["blending"] = kwargs.get('blending') or 'additive'
```
```suggestion
    kwargs['blending'] = kwargs.get('blending') or 'additive'
```

sofroniewn commented on napari/layers/utils/stack_utils.py:
```diff
@@ -0,0 +1,212 @@
+    n_channels = (data[0] if multiscale else data).shape[channel_axis]
+
+    kwargs["blending"] = kwargs.get('blending') or 'additive'
+    kwargs.setdefault("colormap", None)
```
```suggestion
    kwargs.setdefault('colormap', None)
```

sofroniewn reviewed (approved):
This looks great to me too - made a minor suggested changes on `'` vs `"` which i will commit now. Once tests pass again I will merge. Thanks for all the hard work @cwood1967!!

sofroniewn committed:
Apply suggestions from code review
